Projek Bengkel Koding : Hungarian Dataset

Nama                  : Nicholaus Verdhy Putranto

Nim                   : A11.2020.12447

Kelas                 : BKDS07

# Data Collection and Data understanding

Data set pada projek ini diambil dari https://archive.ics.uci.edu/dataset/45/heart+disease.

Direktori ini berisi 4 database mengenai diagnosis penyakit jantung.
   Semua atribut bernilai numerik. Data dikumpulkan dari
   empat lokasi berikut:

     1. Yayasan Klinik Cleveland (cleveland.data)
     2. Institut Kardiologi Hongaria, Budapest (hungarian.data)
     3.V.A. Pusat Medis, Long Beach, CA (long-beach-va.data)
     4. Rumah Sakit Universitas, Zurich, Swiss (switzerland.data)

   Setiap database memiliki format instance yang sama. Sedangkan databasenya ada 76
   atribut mentah, hanya 14 yang benar-benar digunakan. Atribut yang digunakan diantaranya
   
1. #3 (age): usia
2. #4 (sex): Jenis Kelamin
3. #9 (cp): jenis penyakit dada
4. #10 (trestbps): tekanan darah istirahat (dalam mm Hg saat masuk rumah sakit)
5. #12 (chol): kolesterol serum dalam mg/dl
6. #16 (fbs): (gula darah puasa >120 mg/dl)
7. #19 (restecg): hasil elektrokardiografi istirahat
8. #32 (thalach) : detak jantung maksimal tercapai
9. #38 (exang): angina akibat olahraga
10. #40 (oldpeak): Depresi ST yang disebabkan oleh olahraga versus istirahat
11. #41 (kemiringan): kemiringan latihan segmen ST puncak
12. #44 (ca): jumlah bejana besar (0-3) yang diwarnai dengan flourosopy
13. #51 (thal): 3 = biasa; 6 = cacat tetap; 7 = cacat reversibel
14. #58 (num): diagnosis penyakit jantung (status penyakit angiografi)
      
Pada sesi kali ini, saya menggunakan data hungarian.data untuk projek ini

# Load Dataset

Untuk memuat data, kita lakukan import library terlebih dahulu. Import ini ditujukan untuk memasukan pustaka dalam menanangani pemrosesan dalam pengolahan data. 

In [1]:
#Import library 
import pandas as pd
import numpy as np
import re
import itertools

In [2]:
#load dataset
dir = "hungarian.data"

In [3]:
#melakukan encoding pada dataset
with open(dir, encoding='Latin1') as file:
    lines = [line.strip() for line in file]
    
#membuatnya kedalam data frame
data = itertools.takewhile(
    lambda x: len(x) == 76,
    (' '.join(lines[i:(i + 10)]).split() for i in range(0, len(lines), 10)))

df = pd.DataFrame.from_records(data)
df.head()

0  1   2  3  4  5  6   7  8    9   ...  66  67 68 69 70 71 72   73   74  \
0  1254  0  40  1  1  0  0  -9  2  140  ...  -9  -9  1  1  1  1  1  -9.  -9.   
1  1255  0  49  0  1  0  0  -9  3  160  ...  -9  -9  1  1  1  1  1  -9.  -9.   
2  1256  0  37  1  1  0  0  -9  2  130  ...  -9  -9  1  1  1  1  1  -9.  -9.   
3  1257  0  48  0  1  1  1  -9  4  138  ...   2  -9  1  1  1  1  1  -9.  -9.   
4  1258  0  54  1  1  0  1  -9  3  150  ...   1  -9  1  1  1  1  1  -9.  -9.   

     75  
0  name  
1  name  
2  name  
3  name  
4  name  

[5 rows x 76 columns]

In [4]:
# Melihat informasti tipe data yang ada serta jumlah data yang ada setiap vaiabel
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 76 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       294 non-null    object
 1   1       294 non-null    object
 2   2       294 non-null    object
 3   3       294 non-null    object
 4   4       294 non-null    object
 5   5       294 non-null    object
 6   6       294 non-null    object
 7   7       294 non-null    object
 8   8       294 non-null    object
 9   9       294 non-null    object
 10  10      294 non-null    object
 11  11      294 non-null    object
 12  12      294 non-null    object
 13  13      294 non-null    object
 14  14      294 non-null    object
 15  15      294 non-null    object
 16  16      294 non-null    object
 17  17      294 non-null    object
 18  18      294 non-null    object
 19  19      294 non-null    object
 20  20      294 non-null    object
 21  21      294 non-null    object
 22  22      294 non-null    ob

Sesuai dengan arahan pada sumber dataset, terdapat kolom nama 75 berisikan nilai name. Name mewakili ahkir dari setiap baris data, dan nilai -9.0 mewakili missing value. Maka dari itu kita perlu menghapus kolom 75 dan mengubah tipe data object menjadi float agar kita dapat mengubah -9.0 menjadi Nan

In [5]:
#Menghapus kolom 75 berisikan name
df.drop(75,axis=1,inplace=True)

#menhapus kolom 0 karena itu index data
df.drop(0,axis=1,inplace=True)

In [6]:
#mengubah tipe data kolom object menjadi float
df = df.astype(float)

In [7]:
#melihat informasi tipe data 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 74 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   1       294 non-null    float64
 1   2       294 non-null    float64
 2   3       294 non-null    float64
 3   4       294 non-null    float64
 4   5       294 non-null    float64
 5   6       294 non-null    float64
 6   7       294 non-null    float64
 7   8       294 non-null    float64
 8   9       294 non-null    float64
 9   10      294 non-null    float64
 10  11      294 non-null    float64
 11  12      294 non-null    float64
 12  13      294 non-null    float64
 13  14      294 non-null    float64
 14  15      294 non-null    float64
 15  16      294 non-null    float64
 16  17      294 non-null    float64
 17  18      294 non-null    float64
 18  19      294 non-null    float64
 19  20      294 non-null    float64
 20  21      294 non-null    float64
 21  22      294 non-null    float64
 22  23

# Validasi Data

Seperti yang tertera pada sumber data set, -0.9 mewakili missing value. kita gunakan .replace untuk mengganti nilai

In [8]:
df.replace(-9.0,np.nan, inplace=True)

In [9]:
#melihat missing value
df.isna().sum()

1       0
2       0
3       0
4       0
5       0
     ... 
70      0
71      0
72      0
73    266
74    294
Length: 74, dtype: int64

In [10]:
df.isna().sum().sum()

8511

Pada data set ini tertera nilai missing value sebanyak 8511 data

In [11]:
#sekarang mari kita lihat 5 data teratas pada data set. kita gunakan pd.set_option untuk melihat full kolom
pd.set_option("Display.max_column",None)
df.head()

1     2    3    4    5    6   7    8      9    10     11  12  13  14   15  \
0  0.0  40.0  1.0  1.0  0.0  0.0 NaN  2.0  140.0  0.0  289.0 NaN NaN NaN  0.0   
1  0.0  49.0  0.0  1.0  0.0  0.0 NaN  3.0  160.0  1.0  180.0 NaN NaN NaN  0.0   
2  0.0  37.0  1.0  1.0  0.0  0.0 NaN  2.0  130.0  0.0  283.0 NaN NaN NaN  0.0   
3  0.0  48.0  0.0  1.0  1.0  1.0 NaN  4.0  138.0  0.0  214.0 NaN NaN NaN  0.0   
4  0.0  54.0  1.0  1.0  0.0  1.0 NaN  3.0  150.0  0.0    NaN NaN NaN NaN  0.0   

   16  17   18    19    20    21   22   23   24   25   26     27    28   29  \
0 NaN NaN  0.0  12.0  16.0  84.0  0.0  0.0  0.0  0.0  0.0  150.0  18.0  NaN   
1 NaN NaN  0.0  11.0  16.0  84.0  0.0  0.0  0.0  0.0  0.0    NaN  10.0  9.0   
2 NaN NaN  1.0  11.0  21.0  84.0  0.0  0.0  0.0  0.0  0.0  100.0  10.0  NaN   
3 NaN NaN  0.0   9.0  21.0  84.0  0.0  0.0  0.0  0.0  0.0   50.0   5.0  4.0   
4 NaN NaN  0.0   7.0  25.0  84.0  0.0  0.0  1.0  1.0  0.0   25.0   2.0  NaN   

    30     31     32     33     34     35    36   37   38   39   40    41  \
0  7.0  172.0   86.0  200.0  110.0  140.0  86.0  0.0  0.0  0.0  NaN  26.0   
1  7.0  156.0  100.0  220.0  106.0  160.0  90.0  0.0  0.0  1.0  2.0  14.0   
2  5.0   98.0   58.0  180.0  100.0  130.0  80.0  0.0  0.0  0.0  NaN  17.0   
3  4.0  108.0   54.0  210.0  106.0  138.0  86.0  1.0  0.0  1.5  2.0  19.0   
4  3.0  122.0   74.0  130.0  100.0  150.0  90.0  0.0  1.0  0.0  NaN  13.0   

     42  43  44  45  46  47  48  49  50  51  52  53    54    55    56   57  \
0  20.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN  12.0  20.0  84.0  0.0   
1  13.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN  11.0  20.0  84.0  1.0   
2  14.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN  11.0  26.0  84.0  0.0   
3  22.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN   9.0  30.0  84.0  3.0   
4   9.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN   7.0  30.0  84.0  0.0   

   58   59   60  61   62  63  64  65   66  67   68   69   70   71   72  73  74  
0 NaN  NaN  NaN NaN  NaN NaN NaN NaN  NaN NaN  1.0  1.0  1.0  1.0  1.0 NaN NaN  
1 NaN  NaN  2.0 NaN  NaN NaN NaN NaN  NaN NaN  1.0  1.0  1.0  1.0  1.0 NaN NaN  
2 NaN  NaN  NaN NaN  NaN NaN NaN NaN  NaN NaN  1.0  1.0  1.0  1.0  1.0 NaN NaN  
3 NaN  2.0  NaN NaN  2.0 NaN NaN NaN  2.0 NaN  1.0  1.0  1.0  1.0  1.0 NaN NaN  
4 NaN  NaN  NaN NaN  1.0 NaN NaN NaN  1.0 NaN  1.0  1.0  1.0  1.0  1.0 NaN NaN

# Menentukan Fitur

Sesuai dengan sumber penjjelasan datasetnya, fitur yang dugunakan ada pada kolom yang ada pada lokasi [1, 2, 7,8,10,14,17,30,36,38,39,42,49,56]. Maka dari itu kita melakukan pengambilan kolom tersebut

In [12]:
df_selected = df.iloc[:, [1, 2, 7,8,10,14,17,30,36,38,39,42,49,56]]
df_selected

2    3    8      9      11   15   18     31   37   39   40  43   50  \
0    40.0  1.0  2.0  140.0  289.0  0.0  0.0  172.0  0.0  0.0  NaN NaN  NaN   
1    49.0  0.0  3.0  160.0  180.0  0.0  0.0  156.0  0.0  1.0  2.0 NaN  NaN   
2    37.0  1.0  2.0  130.0  283.0  0.0  1.0   98.0  0.0  0.0  NaN NaN  NaN   
3    48.0  0.0  4.0  138.0  214.0  0.0  0.0  108.0  1.0  1.5  2.0 NaN  NaN   
4    54.0  1.0  3.0  150.0    NaN  0.0  0.0  122.0  0.0  0.0  NaN NaN  NaN   
..    ...  ...  ...    ...    ...  ...  ...    ...  ...  ...  ...  ..  ...   
289  48.0  0.0  2.0    NaN  308.0  0.0  1.0    NaN  NaN  2.0  1.0 NaN  NaN   
290  36.0  1.0  2.0  120.0  166.0  0.0  0.0  180.0  0.0  0.0  NaN NaN  NaN   
291  48.0  1.0  3.0  110.0  211.0  0.0  0.0  138.0  0.0  0.0  NaN NaN  6.0   
292  47.0  0.0  2.0  140.0  257.0  0.0  0.0  135.0  0.0  1.0  1.0 NaN  NaN   
293  53.0  1.0  4.0  130.0  182.0  0.0  0.0  148.0  0.0  0.0  NaN NaN  NaN   

      57  
0    0.0  
1    1.0  
2    0.0  
3    3.0  
4    0.0  
..   ...  
289  0.0  
290  0.0  
291  0.0  
292  0.0  
293  0.0  

[294 rows x 14 columns]

Setelah itu, kita mengganti kolom  nama sesuai dengan arahan dari sumber data set. kita gunakan .rename untuk menggantinya. Sebelumnya kita membuat dictionary yang nantinya kita gunakan untuk mengganti nama kolom

Agar tidak menghilangkan data awal / mentah. Mari kita copy dataframenya kedalam df_selected_copy. Hal ini untuk mengantisipasi agar data asli tidak terkontaminasi dengan data yang telah di prosesing

In [13]:
df_selected_copy = df_selected.copy()

In [14]:
#buat dictionary nama
nama_kolom = {
    2: 'age',
    3: 'sex',
    8: 'cp',
    9: 'trestbps',
    11: 'chol',
    15: 'fbs',
    18: 'restecg',
    31: 'thalach',
    37: 'exang',
    39: 'oldpeak',
    40: 'slope',
    43: 'ca',
    50: 'thal',
    57: 'target'
}

In [15]:
#mengubah nama kolom .rename
df_selected_copy.rename(columns=nama_kolom,inplace=True)

In [16]:
df_selected_copy.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  40.0  1.0  2.0     140.0  289.0  0.0      0.0    172.0    0.0      0.0   
1  49.0  0.0  3.0     160.0  180.0  0.0      0.0    156.0    0.0      1.0   
2  37.0  1.0  2.0     130.0  283.0  0.0      1.0     98.0    0.0      0.0   
3  48.0  0.0  4.0     138.0  214.0  0.0      0.0    108.0    1.0      1.5   
4  54.0  1.0  3.0     150.0    NaN  0.0      0.0    122.0    0.0      0.0   

   slope  ca  thal  target  
0    NaN NaN   NaN     0.0  
1    2.0 NaN   NaN     1.0  
2    NaN NaN   NaN     0.0  
3    2.0 NaN   NaN     3.0  
4    NaN NaN   NaN     0.0

# Data Cleansing

Memasuki tahap berikutnya yaitu data cleansing. Data cleansing ini akan berguna untuk menghilangkan missing values pada dataset

In [17]:
# melihat missing value pada kolom
df_selected_copy.isnull().sum()

age           0
sex           0
cp            0
trestbps      1
chol         23
fbs           8
restecg       1
thalach       1
exang         1
oldpeak       0
slope       190
ca          290
thal        266
target        0
dtype: int64

In [18]:
#jumlah data
df_selected_copy.shape

(294, 14)

In [19]:
df_selected_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    float64
 1   sex       294 non-null    float64
 2   cp        294 non-null    float64
 3   trestbps  293 non-null    float64
 4   chol      271 non-null    float64
 5   fbs       286 non-null    float64
 6   restecg   293 non-null    float64
 7   thalach   293 non-null    float64
 8   exang     293 non-null    float64
 9   oldpeak   294 non-null    float64
 10  slope     104 non-null    float64
 11  ca        4 non-null      float64
 12  thal      28 non-null     float64
 13  target    294 non-null    float64
dtypes: float64(14)
memory usage: 32.3 KB


Kita bisa lihat pada missing value, pada kolom slope,ca, dan thal memiliki banyak nilai yang hilang hampir 90%, maka kita hapus kolom tersebut terlebih dahulu

In [20]:
df_selected_copy.drop(["slope","ca","thal"],axis=1, inplace=True)

In [21]:
df_selected_copy.isna().sum()

age          0
sex          0
cp           0
trestbps     1
chol        23
fbs          8
restecg      1
thalach      1
exang        1
oldpeak      0
target       0
dtype: int64

Setelah itu mari kita isi nilai missing value tersebut dengan nilai mean

In [22]:
df_selected_copy.describe()

age         sex          cp    trestbps        chol         fbs  \
count  294.000000  294.000000  294.000000  293.000000  271.000000  286.000000   
mean    47.826531    0.724490    2.982993  132.583618  250.848708    0.069930   
std      7.811812    0.447533    0.965117   17.626568   67.657711    0.255476   
min     28.000000    0.000000    1.000000   92.000000   85.000000    0.000000   
25%     42.000000    0.000000    2.000000  120.000000  209.000000    0.000000   
50%     49.000000    1.000000    3.000000  130.000000  243.000000    0.000000   
75%     54.000000    1.000000    4.000000  140.000000  282.500000    0.000000   
max     66.000000    1.000000    4.000000  200.000000  603.000000    1.000000   

          restecg     thalach       exang     oldpeak      target  
count  293.000000  293.000000  293.000000  294.000000  294.000000  
mean     0.218430  139.129693    0.303754    0.586054    0.792517  
std      0.460868   23.589749    0.460665    0.908648    1.237006  
min      0.000000   82.000000    0.000000    0.000000    0.000000  
25%      0.000000  122.000000    0.000000    0.000000    0.000000  
50%      0.000000  140.000000    0.000000    0.000000    0.000000  
75%      0.000000  155.000000    1.000000    1.000000    1.000000  
max      2.000000  190.000000    1.000000    5.000000    4.000000

Nilai mean dari kolom yang memiliki missing value sudah bertipe float. Kita bisa langsung mengisi nilai hilang dengan fillna dan .mean

In [23]:
df_selected_copy.fillna(df_selected_copy.mean(),inplace=True)
df_selected_copy.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
target      0
dtype: int64

In [24]:
df_selected_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    float64
 1   sex       294 non-null    float64
 2   cp        294 non-null    float64
 3   trestbps  294 non-null    float64
 4   chol      294 non-null    float64
 5   fbs       294 non-null    float64
 6   restecg   294 non-null    float64
 7   thalach   294 non-null    float64
 8   exang     294 non-null    float64
 9   oldpeak   294 non-null    float64
 10  target    294 non-null    float64
dtypes: float64(11)
memory usage: 25.4 KB


Setelah itu saya melakukan pengecekan pada duplikasi data dengan .duplicate

In [25]:
df_selected_copy.duplicated().sum()

1

Terdapat 1 data duplikat, mari kita lihat dalam dataset

In [26]:
df_duplikat = df_selected_copy.duplicated()
df_selected_copy[df_selected_copy.duplicated(keep=False)]

age  sex   cp  trestbps        chol  fbs  restecg  thalach  exang  \
90   49.0  0.0  2.0     110.0  250.848708  0.0      0.0    160.0    0.0   
163  49.0  0.0  2.0     110.0  250.848708  0.0      0.0    160.0    0.0   

     oldpeak  target  
90       0.0     0.0  
163      0.0     0.0

setelah kita mgengetahui ada duplikasi data, mari kita hapus duplikat datanya

In [27]:
df_selected_copy.drop_duplicates(inplace=True)

In [28]:
df_selected_copy.duplicated().sum()

0